api/envoy/extensions/filters/http/local_ratelimit/v3/local_rate_limit.proto

```diff
@@ -110,6 +110,19 @@ message LocalRateLimit {
   // If unspecified, the default value is false.
   bool local_rate_limit_per_downstream_connection = 11;
 
+  // Enables the local cluster level rate limiting, iff this is set explicitly.
+  //
+  // .. note::
+  //   This should never be set if the ``local_rate_limit_per_downstream_connection`` is set to
+  //   true. Because if per connection rate limiting is enabled, we assume that the token buckets
+  //   should never be shared across Envoy instances.
+  //
+  // .. note::
+  //   This only works when the :ref:`local cluster name
+  //   <envoy_v3_api_field_config.bootstrap.v3.ClusterManager.local_cluster_name>` is set and
+  //   related cluster is defined in the bootstrap configuration.
+  common.ratelimit.v3.LocalClusterRateLimit local_cluster_rate_limit = 16;
```

Comment:
High-level comments:
I've read through the comments in #34230 and if I understand the intent correctly (feel free to correct me here), the motivation is to have tokens that are dependent on the number of endpoints a cluster has.
1. While I'm not a fan of interleaving the routing and the load-balancing parts, I can see where this can be beneficial. However, it seems to me that the goal should probably be, reducing/blocking the queries to those instances that are busy. Will using [ClientSideWeightedRoundRobin](https://github.com/envoyproxy/envoy/blob/c1c1619aec370fda1bf805955b2e460f488c29ab/api/envoy/extensions/load_balancing_policies/client_side_weighted_round_robin/v3/client_side_weighted_round_robin.proto#L38C9-L38C37) make more sense? the benefit is that it is more dynamic then the static allocation, and allows keeping the logic in the load-balancing part of Envoy. (FWIW, we are currently working on this implementation).
2. If the idea is to have X tokens per instance, so the cluster has N*X tokens (where N is the number of endpoints), why restrict this to Local-Clusters only?
3. The work refers to Local-Cluster which AFAIK is the cluster where the Envoy instance is the sidecar. The description in #34230 implies that this work allows sharing the info between the different Envoy instances. It is unclear to me why the number of Envoy instances is important here. IIUC this is going to have the same impact as the current local-rate-limit. The only difference I'm seeing here is that instead of having a strict tokens number per cluster, there are going to be strict tokens number per endpoint of the cluster. Is this the intent? if so, lets simplify by not focusing on multiple Envoy instances. WDYT?

Reply:
Sorry, I didn't make it clear. This is most design for gateway mode. Given a specific example may be better:

Assume we have an Envoy gateway with 3 instances Envoy. And an user configured a local rate limit rule with 3000 tokens/second for route A. Because the local rate limit filter works in independently in different instances, the actual throughput is 3000*3=9000 tokens/second.

Now, assume we change the replica number of the gateway to 4, for some reasons. Then actual throughput of route A  becomes to 3000*4=12000 tokens/second.

In most of cases, this is unexpected change for end users. Because not every user know the implementation detail of the rate limit filter. They may expect a stable throughput for the whole gateway, regardless of the number of replica.
For example, given a gateway and a specific route, if X tokens/second is set for the route, then X/N tokens/second should be used by every single Envoy instance, where the N is the number of Envoy replica. By this way, no matter how much the Envoy replica number is, the final throughput of the route A would be X tokens/second.

The PR introduced the `local_cluster_rate_limit` to let the rate limit filter to aware the replica number of the local cluster (the cluster contains current Envoy instance self), then we can calculate the X/N dynamically.

cc @adisuissa

Reply:
I think I may be misunderstanding the definition of `local_cluster_name`.
My understanding was that it will set the cluster name that this Envoy is adjacent to, not that its endpoints need to be the Envoys that comprise the cluster.
Can you please point me to where you see that it needs to contain the Envoy hosts that comprise the cluster?

Reply:
> ([string](https://developers.google.com/protocol-buffers/docs/proto#scalar)) Name of the local cluster (i.e., the cluster that owns the Envoy running this configuration). In order to enable [zone aware routing](https://www.envoyproxy.io/docs/envoy/latest/intro/arch_overview/upstream/load_balancing/zone_aware#arch-overview-load-balancing-zone-aware-routing) this option must be set. If local_cluster_name is defined then [clusters](https://www.envoyproxy.io/docs/envoy/latest/api-v3/config/cluster/v3/cluster.proto#envoy-v3-api-msg-config-cluster-v3-cluster) must be defined in the [Bootstrap static cluster resources](https://www.envoyproxy.io/docs/envoy/latest/api-v3/config/bootstrap/v3/bootstrap.proto#envoy-v3-api-field-config-bootstrap-v3-bootstrap-staticresources-clusters). This is unrelated to the [--service-cluster](https://www.envoyproxy.io/docs/envoy/latest/operations/cli#cmdoption-service-cluster) option which does not [affect zone aware routing](https://github.com/envoyproxy/envoy/issues/774).

https://www.envoyproxy.io/docs/envoy/latest/api-v3/config/bootstrap/v3/bootstrap.proto#envoy-v3-api-msg-config-bootstrap-v3-clustermanager

In the gateway mode, the local cluster should be cluster that composed by current Envoy instance and its siblings.
In the sidecar mode, the local cluster should be cluster that composed by current app Instance where the envoy sidecar is running and its siblings.

Reply:
Apologies for the delayed review.
I read the descriptions, and looked at the PRs (from 2016) that added the field.
While I agree that the sidecar mode makes sense, it is unclear to me why for Gateway "the cluster that owns the Envoy running this configuration" dictates that the cluster needs to be comprised of all the Envoy instances (and not some cluster that is "local"/"zone" that this Envoy is associated with, and should forward requests to that cluster in the zone). If the case is as you describe, I think it should be clarified (possibly with some example for that field).

Nevertheless, I still think that having "Envoy's left-brain" (listener/filter) be influenced by the properties of "Envoy's right-brain" (cluster/endpoints) is not desired. More specifically, what should happen to the number of tokens if some instances are unhealthy, or have different weights, what should be done if multiple priorities are used? If, in the future, one would want to support these cases, the left/right brain model will break.

Maybe I'll ask it differently, why does it matter if the number of tokens is modified by an EDS or an ECDS update?

Reply:
> While I agree that the sidecar mode makes sense, it is unclear to me why for Gateway "the cluster that owns the Envoy running this configuration" dictates that the cluster needs to be comprised of all the Envoy instances (and not some cluster that is "local"/"zone" that this Envoy is associated with, and should forward requests to that cluster in the zone). If the case is as you describe, I think it should be clarified (possibly with some example for that field).

The local cluster not must be the cluster that comprised by all these gateway Envoy instances. There is no any strict validation to that. But it should be. As I said in previous comments, the `local cluster` is mainly used to tell the distribution of Envoy instances in different zone. Then Envoy could determine how to route the traffic to upstream cluster fairly.
For example, given an edge gateway that distribute in three different zone, it's necessary to tell the gateway how much Envoy instances in current zone. Then, Envoy could know how much traffic should be routed to current zone and how much should be routed to other zone.


> Nevertheless, I still think that having "Envoy's left-brain" (listener/filter) be influenced by the properties of "Envoy's right-brain" (cluster/endpoints) is not desired. More specifically, what should happen to the number of tokens if some instances are unhealthy, or have different weights, what should be done if multiple priorities are used? If, in the future, one would want to support these cases, the left/right brain model will break.

It's hard to say the local cluster is Envoy's right-brain. The local cluster never take any traffic from Envoy (it's originating cluster, is the Envoy self). We just reuse the cluster/EDS here to tell Envoy about it's siblings' number and distribution.
If someone want to support weighted token share, than new field should be added in the `common.ratelimit.v3.LocalClusterRateLimit` message to enable it. (This is why I define a message rather than just using a bool.)

> Maybe I'll ask it differently, why does it matter if the number of tokens is modified by an EDS or an ECDS update?

Note, the tokens is not modified by an EDS update. It's changed with the gateway instance number's change. EDS only is a way to tell the gateway instance number is changed.

By this way, for example, for specific route, we can ensure the total tokens for the whole gateway will never be changed, even the gateway instances is changed continuously. This is my target: stable total tokens of local rate limit filter for whole gateway.

Reply:
I think more docs (rst files) could be beneficial in explaining these cases. This can be in a future PR.